semantics/src/main/scala/aqua/semantics/expr/func/ForSem.scala

```diff
@@ -75,18 +78,39 @@ class ForSem[S[_]](val expr: ForExpr[S]) extends AnyVal {
 
 object ForSem {
 
+  def itemOrPair[S[_]](nameOrPair: ForExpr.NameOrPair[S]): (String, Option[ForKeyValue]) =
+    nameOrPair match {
+      case Right(v) => (v.value, None)
+      case Left(k, v) => ("-iterable-", ForKeyValue(k.value, v.value).some)
+    }
+
   def beforeFor[S[_], F[_]: Monad](
-    item: Name[S],
+    item: ForExpr.NameOrPair[S],
     iterable: ValueToken[S]
   )(using
     V: ValuesAlgebra[S, F],
     N: NamesAlgebra[S, F],
-    T: TypesAlgebra[S, F]
+    M: ManglerAlgebra[F]
   ): F[Option[ValueRaw]] = (for {
     value <- V.valueToIterable(iterable)
     (raw, typ) = value
-    _ <- OptionT.liftF(
-      N.define(item, typ.element)
-    )
-  } yield raw).value
+    res <- (typ, item) match {
+      case (smt: StreamMapType, Left(key, value)) =>
+        OptionT.liftF(for {
+          _ <- N.define(key, ScalarType.string)
+          _ <- N.define(value, smt.element)
+        } yield raw)
+      case (smt: StreamMapType, Right(it)) =>
+        val typeName = "-streamMapIter-"
+        OptionT.liftF(for {
+          newTypeName <- M.rename(typeName)
+          iterType = smt.iterType(newTypeName)
+          _ <- N.define(it, iterType)
+        } yield raw)
+      case (_, Left(_, _)) =>
+        OptionT.none
```

Comment:
Why just `none`? Why not returning a semantic error?
This code just silently compiles:
```aqua
aqua Main

export main

service Test("test-srv"):
  call()

func main():
  stream: *string
  for k, v <- stream:
    Test.call()
```